eclair-core/src/main/scala/fr/acinq/eclair/channel/InteractiveTxBuilder.scala

```diff
@@ -0,0 +1,721 @@
+    log.debug("contributing {} to interactive-tx construction", toFund)
+    if (toFund <= 0.sat) {
+      // We're not the initiator and we don't want to contribute to the funding transaction.
+      buildTx(FundingContributions(Nil, Nil))
+    } else {
+      // We always double-spend all our previous inputs.
+      val previousInputs = previousAttempts.flatMap(_.tx.localInputs).distinctBy(_.serialId)
+      val dummyTx = Transaction(2, previousInputs.map(i => TxIn(toOutPoint(i), ByteVector.empty, i.sequence)), Seq(TxOut(toFund, fundingParams.fundingPubkeyScript)), fundingParams.lockTime)
+      fund(dummyTx, previousInputs, Set.empty)
+    }
+  }
+
+  def fund(txNotFunded: Transaction, currentInputs: Seq[TxAddInput], unusableInputs: Set[OutPoint]): Behavior[Command] = {
+    context.pipeToSelf(wallet.fundTransaction(txNotFunded, fundingParams.targetFeerate, replaceable = true, lockUtxos = true)) {
+      case Failure(t) => WalletFailure(t)
+      case Success(result) => FundTransactionResult(result.tx)
+    }
+    Behaviors.receiveMessagePartial {
+      case FundTransactionResult(fundedTx) =>
+        filterInputs(fundedTx, currentInputs, unusableInputs)
+      case WalletFailure(t) =>
+        log.error("could not fund dual-funded channel: ", t)
+        // We use a generic exception and don't send the internal error to the peer.
+        replyTo ! LocalFailure(ChannelFundingError(fundingParams.channelId))
+        unlockAndStop(currentInputs.map(toOutPoint).toSet ++ unusableInputs)
+      case msg: ReceiveMessage =>
+        timers.startSingleTimer(msg, 1 second)
+        Behaviors.same
+      case Abort =>
+        timers.startSingleTimer(Abort, 1 second)
+        Behaviors.same
+    }
+  }
+
+  def filterInputs(fundedTx: Transaction, currentInputs: Seq[TxAddInput], unusableInputs: Set[OutPoint]): Behavior[Command] = {
+    context.pipeToSelf(Future.sequence(fundedTx.txIn.map(txIn => getInputDetails(txIn, currentInputs)))) {
+      case Failure(t) => WalletFailure(t)
+      case Success(results) => InputDetails(results.collect { case Right(i) => i }, results.collect { case Left(i) => i }.toSet)
+    }
+    Behaviors.receiveMessagePartial {
+      case inputDetails: InputDetails =>
+        if (inputDetails.unusableInputs.isEmpty) {
+          // This funding iteration did not add any unusable inputs, so we can directly return the results.
+          val changeOutputs = fundedTx.txOut
+            .filter(_.publicKeyScript != fundingParams.fundingPubkeyScript)
+            .map(txOut => TxAddOutput(fundingParams.channelId, generateSerialId(), txOut.amount, txOut.publicKeyScript))
+          val outputs = if (fundingParams.isInitiator) {
+            // If the initiator doesn't want to contribute, we should cancel out the dust amount artificially added previously.
+            val initiatorChangeOutputs = if (fundingParams.localAmount == 0.sat) {
+              changeOutputs.map(o => o.copy(amount = o.amount + fundingParams.dustLimit))
```

Comment:
There will always be a single change output, we are getting a list because that's what `filter` returns, but `bitcoind` never adds more than one change output. Do you think I should explicitly only update the first one? It's a bit more code but it's true that it would remove an implicit subtlety.

Reply:
I added a log line in https://github.com/ACINQ/eclair/pull/2273/commits/83e51e46c9bf4e6617b7c7d77ce2af24da41d6de if that isn't the case. I think it's better than a `require`, since it would be overkill to fail the dual funding attempt for this, we'll just end up paying a bit less fees if that assumption is incorrect which is probably ok.

Reply:
Since both of you were in favor of a `require`, I implemented it in https://github.com/ACINQ/eclair/pull/2273/commits/b3ab1d395f0e0027a23b185c1a36895a458f81aa

Reply:
Maybe we could check that the outputs of the funded tx are as expected, and gracefully `unlockAndStop` if not? Putting a `require` on something that depends on an external software which we can't deterministically test is a little footgunny.

Reply:
Good idea, I like that better. All cases are explicitly handled in https://github.com/ACINQ/eclair/pull/2273/commits/d5c640b75c5a12f04bdd795bdf9d25aa407cc9df